### ISSUE TITLE
Unexpected natural sort when sorting multi-dimensional arrays or `pandas.DataFrame`

### ISSUE THREAD

**Issue by aallahyar** [SethMMorton/natsort#172]
**Describe the bug**
I am seeing an ordered output of `natsort` that does not seem to be right. 

**Expected behavior**
An example can clarify the issue, see below:

```python
import pandas as pd
from natsort import natsort_keygen

df = pd.DataFrame(
    data=range(5), 
    index=["W0", "W", "W54", "W30", "W20"], 
    columns=['col1']
)
print(df)
#      col1
# W0      0
# W       1
# W54     2
# W30     3
# W20     4

```
Now if we sort the indices as a `list`:
```python
sorted(df.index, key=natsort_keygen())

# we get:
# ['W', 'W0', 'W20', 'W30', 'W54']
```
which is expected.

**The issue**
However, if I try to reorder the rows by this:

```python
df.sort_index(key=natsort_keygen())

# we get:
#      col1
# W0      0
# W       1
# W54     2
# W30     3
# W20     4
```

Note that the following works as expected:
```python
import numpy as np
from natsort import index_natsorted

print(df.sort_index(key=lambda x: np.argsort(index_natsorted(df.index))))
#      col1
# W       1
# W0      0
# W20     4
# W30     3
# W54     2
```

**Environment:**
 - Python Version: Python 3.9.18
 - OS: Linux CentOS [e.g. Windows, Fedora]
 - If the bug involves `LOCALE` or `humansorted`: It does not seem to be related to `humansorted`

**To Reproduce**
See above for a reproducible example.

**Comment by aallahyar** [comment]
I did further investigation and the issue seems to be related to how `Pandas` interpret the output of `natsort_keygen()`.

In particular, we get a `tuple` of values from `natsort_keygen()`:
```python
natsort_keygen()(["W0", "W", "W54", "W30", "W20"])

# outputs
# (('W', 0), ('W',), ('W', 54), ('W', 30), ('W', 20))
```

Which [according to the source](https://github.com/pandas-dev/pandas/blob/e14a9bd41d8cd8ac52c5c958b735623fe0eae064/pandas/core/sorting.py#L579), is then fed to `Indexer` to be converted to a `Pandas.Index`. If I try to simulate that, we have
```python
idx = natsort_keygen()(["W0", "W", "W54", "W30", "W20"])
pd.Index(idx)

# we get:
# MultiIndex([('W',),
#             ('W',),
#             ('W',),
#             ('W',),
#             ('W',)],
#            )
```

It does seem to be the root of the issue, and not something that `natsort` does wrong.

Out of curiosity, do you think this can be resolved on `natsort` side?

**Comment by SethMMorton** [comment]
Because of the way `natsort` works, each element does not know about the other elements so it's not really possible to pad-fill missing elements. But, you could do it yourself x(proof-of-concept, not optimized):

```py
key = natsort_keygen()
keyed = [key(x) for x in df.index]
longest = max(keyed, key=len)
to_fill = [len(x) - longest for x in keyed]
pad_filled_keyed = [(*x, [None]*n) for x, n in zip(keyed, to_fill)]
index = pd.Index(pad_filled_keyed)
```

**Comment by aallahyar** [comment]
Thanks @SethMMorton very useful :)
Can this be an argument in your functions? E.g., `keygen(pad=True)` which would activate a "post processing" step to pad fields? This argument could be `False` by default for backward compatibility and speed.
Would you agree with this?

**Comment by SethMMorton** [comment]
No, it's not possible. A sorting key is only passed a single argument, not the whole collection, so each call knows nothing about the other arguments so padding is not possible.

Even if it were possible. padding the results with `None` would render the output unsortable in most contexts:

```py
>>> data = (('W', 0), ('W', None), ('W', 54), ('W', 30), ('W', 20))
>>> sorted(data)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'NoneType' and 'int'
```

`pandas` must have special handling for `nan` that allows this to work.

**Comment by SethMMorton** [comment]
Given that `df.sort_index(key=lambda x: np.argsort(index_natsorted(df.index)))` works, it's ot clear why additional functionality needed - can you help me understand?

**Comment by aallahyar** [comment]
Thanks for the remark, understood!
I also agree with your `index_natsorted` feedback and its correct functionality.

Thanks for this package, its awesome :)

**Comment by aallahyar** [comment]
I have further information about it.

It turns out that the issue also exists with `index_natsorted()` when combined with `pandas.DataFrame` when performing multi-column sorting. An example can clarify:

```python
import numpy as np
import pandas as pd
from natsort import index_natsorted

df = pd.DataFrame({
    'col1': ['A', 'B', 'A', 'A', 'A'],
    'col2': ['W1', 'W3', 'W11', 'W17', 'W2'],
})

df.sort_values(by=['col1', 'col2'])
# output:
#   col1 col2
# 0    A   W1
# 2    A  W11
# 3    A  W17
# 4    A   W2
# 1    B   W3

df.sort_values(by=['col1', 'col2'], key=lambda x: np.argsort(index_natsorted(x)))
# output:
#   col1 col2
# 0    A   W1
# 2    A  W11
# 3    A  W17
# 4    A   W2
# 1    B   W3
```

If you look at `pandas` code [for sorting](https://github.com/pandas-dev/pandas/blob/89b286a699b2d023b7a1ebc468abf230d84ad547/pandas/core/sorting.py#L346), you can see that each `column` is "mapped" independently and then the lexical sorting is performed.
Now, if we simulate this with each column of our dataframe, we get:

```python
print(index_natsorted(df['col1']))
# [0, 2, 3, 4, 1]

print(index_natsorted(df['col2']))
# [0, 4, 1, 2, 3]
```
Notice how the first 4 elements in `col1` are given another index. Now, if we do a lexical sort (e.g., in `numpy`), the sorting would not be correct.

So in conclusion, having different indices for identical values works perfectly when a single-column sorting is intended, but would not work if a multi-column sorting is performed.
I wonder if its possible `index_natsorted()` could produces the same index with identical elements?

Thanks!

**Comment by SethMMorton** [comment]
I am able to get the results you desire using available functionality:

```py
>>> df.sort_values(key=natsort.natsort_keygen(), by=["col1", "col2"]) 
  col1 col2
0    A   W1
4    A   W2
2    A  W11
3    A  W17
1    B   W3
```

**Comment by SethMMorton** [comment]
Actually, I just re-read https://stackoverflow.com/questions/29580978/naturally-sorting-pandas-dataframe/ and remembered this solution.

```py
df.reindex(index=natsorted(df.index))
```

For sorting indexes, this seems pretty clean.

**Comment by SethMMorton** [comment]
Here's something else I just found, which I think is even cleaner:

```py
df.iloc[index_natsorted(df.index)]
```

**Comment by aallahyar** [comment]
Very good! You gave me a nice path to follow. Thanks :)
I was writing the same example in addition to:

```python
df.loc[natsorted(df.index)]
```

Thanks for helping out. This issue can be set to "resolved". :)

**Comment by abubelinha** [comment]
> I am going to add your `sort_index` solution to [this section on how to sort pandas dataframes](https://github.com/SethMMorton/natsort/wiki/Examples-and-Recipes#sorting-a-pandas-dataframe) so that it is complete.

Just out of curiosity, why is the code underlined in that wiki?  I am not sure it is my browser or something intentional.

Thank you both for this illustrative discussion

**Comment by SethMMorton** [comment]
There was a problem with rendering the code-block. I have fixed it - thanks for pointing it out.
